include/sleipnir/control/OCPSolver.hpp

```diff
@@ -0,0 +1,205 @@
+   * Build an optimization problem using a system evolution function (explicit
+   * ODE or discrete state transition function).
+   *
+   * @param numStates The number of system states.
+   * @param numInputs The number of system inputs.
+   * @param dt The timestep for integration. Must be in seconds.
+   * @param numSteps The number of control points.
+   * @param dynamics The system evolution function, either an explicit ODE or a
+   * discrete state transition function.
+   * @param dynamicsType The type of system evolution function.
+   * @param method The transcription method.
+   */
+  FixedStepOCPSolver(int numStates, int numInputs, double dt, int numSteps,
+                     const DynamicsFunction& dynamics,
+                     DynamicsType dynamicsType = kExplicitODE,
+                     TranscriptionMethod method = kDirectTranscription);
+
+  /**
+   * Utility function to constrain the initial state.
+   *
+   * @param initialState the initial state to constrain to.
+   */
+  void ConstrainInitialState(const VariableMatrix& initialState);
+
+  /**
+   * Utility function to constrain the final state.
+   *
+   * @param initialState the final state to constrain to.
+   */
+
+  void ConstrainFinalState(const VariableMatrix& finalState);
+
+  /**
+   * Set the constraint evaluation function. This function is called
+   * `numSteps+1` times, with the corresponding state and input
+   * VariableMatrices.
+   *
+   * @param constraintFunction the constraint function.
+   */
+  void ConstrainAlways(const FixedStepConstraintFunction& constraintFunction);
+
+  /**
+   * Convenience function to set a lower bound on the input.
+   *
+   * @param lowerBound The lower bound that inputs must always be above. Must be
+   * shaped (numInputs)x1.
+   */
+  void SetLowerInputBound(const VariableMatrix& lowerBound) {
+    SubjectTo(U() >= lowerBound);
+  }
+
+  /**
+   * Convenience function to set an upper bound on the input.
+   *
+   * @param lowerBound The upper bound that inputs must always be below. Must be
+   * shaped (numInputs)x1.
+   */
+  void SetUpperInputBound(const VariableMatrix& upperBound) {
+    SubjectTo(U() <= upperBound);
+  }
+
+  /**
+   * Get the state variables. After the problem is solved, this will contain the
+   * optimized trajectory.
+   *
+   * Shaped (numStates)x(numSteps+1).
+   *
+   * @returns The state variable matrix.
+   */
+  VariableMatrix& X() { return m_X; };
+
+  /**
+   * Get the input variables. After the problem is solved, this will contain the
+   * inputs corresponding to the optimized trajectory.
+   *
+   * Shaped (numInputs)x(numSteps+1), although the last input step is unused in
+   * the trajectory.
+   *
+   * @returns The input variable matrix.
+   */
+  VariableMatrix& U() { return m_U; };
+
+  /**
+   * Convenience function to get the initial state in the trajectory.
+   *
+   * @returns The initial state of the trajectory.
+   */
+  VariableMatrix InitialState() { return m_X.Col(0); }
```

Comment:
Should this function be const-qualified and return const ref?

Reply:
I don't think this can be, because `Col` doesn't return a const ref.

Reply:
Col() returns a `VariableBlock<VariableMatrix>`, which is a reference to the original autodiff variable. Casting to VariableMatrix makes a copy, so whatever the user does with it won't have any effect. The return type should be changed to `VariableBlock<VariableMatrix>`.

Reply:
Doesn't the use of intrusive shared pointers make even copies essentially work the same as references?

Reply:
Theoretically yes? A unit test should be written to cover these functions.

Reply:
Most operations return a new Variable, so assigning the result back to the copy won't affect the original.